### PR TITLE
Implement ConstructPortalEvent

### LINKED ISSUE #1559: Implementation Request: ConstructPortalEvent
This event doesn't seem to be implemented as it doesn't fire when I build a portal. Could you please implement it? Thanks a lot.

Edit: It doesn't fire when travelling through a portal creates another exit portal either.

### PR DESCRIPTION
**Pull request by JBYoshi**
Supports:
- [x] Lighting nether portals
- [x] Nether portals created after teleportation
- [x] Inserting ender eyes into End portals
- [ ] Spawning the End exit portal

Fixes #1559.

### CODE REVIEW

File: testplugins/src/main/java/org/spongepowered/test/PortalTest.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
+package org.spongepowered.test;
```
Comment (kashike) [inline testplugins/src/main/java/org/spongepowered/test/PortalTest.java:25]: license

**Review by kashike — CHANGES_REQUESTED** [review]

File: testplugins/src/main/java/org/spongepowered/test/PortalTest.java (hunk 1 above)
Comment (JBYoshi) [reply testplugins/src/main/java/org/spongepowered/test/PortalTest.java:25]: Fixed.

**Comment by ryantheleach** [comment]
Consider adding support not only for the end exit portal, but end gateways.

**Comment by Zidane** [comment]
@JBYoshi 

Status update?

**Comment by JBYoshi** [comment]
@Zidane I found out that implementing this is going to be somewhat more difficult than I thought. My current approach for detecting portal constructions during teleports (by assuming all block placements during a teleport are part of a new portal) is unreliable. On the one hand, it doesn't work for detecting portals created for the first time. On the other hand, it improperly detects the [obsidian platform](https://minecraft.gamepedia.com/Generated_structures#Obsidian_platform) created when entering the End as a portal. I'll probably try to detect whether a block is a portal, but I don't know of a simple way to do that.

**Comment by ryantheleach** [comment]
@JBYoshi I'd consider that platform as part of the portal personally, it's just the receiving end, Just like nether portals have a base when they are forced into air, the one way only exit of the end portal has a set platform in the end at a specific place in the end. As long as it's documented it's fine IMO, unless you can think of some good reasons why we would want to not include blocks created when a portal spawns in the events?

The main usecase I see, is people wanting to prevent all of the blocks being placed / created, so this is fine, maybe some docs or renaming of the method is required to make it more clear.

File: testplugins/src/main/java/org/spongepowered/test/PortalTest.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    private static final Random RANDOM = new Random();
+    /**
+     * There is no significance in this UUID. I just created it so I didn't
+     * have to go through the effort of setting up custom data.
+     */
+    private static final UUID CREATOR_UUID = UUID.fromString("7fac20d9-cbd4-48c7-88dc-61a33e22cd74");
+
+    @Listener
+    public void onCreatePortal(ConstructPortalEvent e) {
+        if (RANDOM.nextInt(5) == 0) {
```
Comment (gabizou) [inline testplugins/src/main/java/org/spongepowered/test/PortalTest.java]: Make this a toggle as well, I don't want to randomly be testing portals and wondering what this new mechanic where they suddenly explode :P

File: src/main/java/org/spongepowered/common/mixin/core/item/MixinItemEnderEye.java [hunk 3]
```diff
@@ -47,4 +58,26 @@ private boolean onspawnEntity(World world, Entity enderEye, World worldIn, Entit
...
+            "Lnet/minecraft/block/state/pattern/BlockPattern;match(Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;)Lnet/minecraft/block/state/pattern/BlockPattern$PatternHelper;"))
+    @Nullable
+    private BlockPattern.PatternHelper onConstructPortal(BlockPattern pattern, World worldIn, BlockPos pos) {
+        BlockPattern.PatternHelper match = pattern.match(worldIn, pos);
+        if (match == null) {
+            return null;
+        }
+        Vector3d center = VecHelper.toVector3d(match.getFrontTopLeft())
+                .sub(pattern.getThumbLength() / 2.0, pattern.getFingerLength() / 2.0, pattern.getPalmLength() / 2.0);
+        ConstructPortalEvent event = SpongeEventFactory.createConstructPortalEvent(Sponge.getCauseStackManager().getCurrentCause(),
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/item/MixinItemEnderEye.java]: Would we be able to add the blocks being changed for the portal?

File: src/main/java/org/spongepowered/common/entity/EntityUtil.java [hunk 4]
```diff
@@ -346,6 +361,17 @@ public static EntityPlayerMP transferPlayerToWorld(final EntityPlayerMP player,
                 // TODO determine what it truly is for API 8
                 frame.addContext(EventContextKeys.TELEPORT_TYPE, TeleportTypes.PORTAL);
 
+                if (!context.getCapturedBlockChanges().isEmpty()) {
+                    ConstructPortalEvent constructPortalEvent = SpongeEventFactory.createConstructPortalEvent(Sponge.getCauseStackManager()
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/entity/EntityUtil.java]: Can shortcut with `frame.getCurrentcause()` since we are already in a frame.

**Comment by Zidane** [comment]
@JBYoshi Either update this or close it.

**Comment by JBYoshi** [comment]
I'm going to go ahead and close this. This code seems to have broken since I last looked at it, and reimplementing it seems to be much more trouble than it's worth. I'm thinking of just getting rid of `ConstructPortalEvent` entirely. `ChangeBlockEvent` should suffice for most cases, and it provides much more information than `ConstructPortalEvent` could.

**Pull request closed without merging**